### PR TITLE
Previous manuscripts

### PR DESCRIPTION
**Pull request by BalteanuAndrei709**

### CODE REVIEW

File: Backend/src/main/java/ro/uaic/info/romandec/repository/ManuscriptRepository.java [hunk 1]
```diff
@@ -12,4 +16,8 @@ public interface ManuscriptRepository extends JpaRepository<Manuscript,Integer>
     String getRandomNotDecipheredManuscript();
 
     Manuscript getManuscriptByPathToImage(String pathToOriginalImage);
```
Comment (stefan-nistor) [inline Backend/src/main/java/ro/uaic/info/romandec/repository/ManuscriptRepository.java:18]: Change the return type to `Optional<Mannuscript>`

File: Backend/src/main/java/ro/uaic/info/romandec/services/ManuscriptService.java [hunk 2]
```diff
@@ -2,44 +2,168 @@
...
+            return filename;
+        }
+
+        return filename.substring(0, dotIndex);
+    }
+
+    private String getDecipheredTranscriptFilename(String filename)
+    {
+        return filename.split("\\.")[0] + ".txt";
+    }
```
Comment (stefan-nistor) [inline Backend/src/main/java/ro/uaic/info/romandec/services/ManuscriptService.java:68]: Maybe refactor this to reuse the code
```suggestion
    private String getDecipheredTranscriptFilename(String filename) {
    String filenameWithoutExtension = extractFilenameWithoutExtension(filename);
    return filenameWithoutExtension + ".txt";
}
```

File: Backend/src/main/java/ro/uaic/info/romandec/services/ManuscriptService.java [hunk 3]
```diff
@@ -2,44 +2,168 @@
...
+
+        int dotIndex = filename.lastIndexOf('.');
+
+        if (dotIndex <= 0)
+        {
+            return filename;
+        }
+
+        return filename.substring(0, dotIndex);
+    }
```
Comment (stefan-nistor) [inline Backend/src/main/java/ro/uaic/info/romandec/services/ManuscriptService.java:63]: Maybe make use of Java Paths
```suggestion
    private String extractFilenameWithoutExtension(String filename) {
    Path path = Paths.get(filename);
    return path.getName(path.getNameCount() - 1).toString();
}
```

File: Backend/src/main/java/ro/uaic/info/romandec/services/ManuscriptService.java [hunk 4]
```diff
@@ -2,44 +2,168 @@
...
+        Optional<Manuscript> manuscript = manuscriptRepository.getManuscriptByIdAndNameAndUserId(
+                request.getManuscriptId(),
+                request.getName(),
+                userId);
+
+        if (manuscript.isEmpty()){
+            throw new NoAvailableDataForGivenInput("No manuscript found for id:" + manuscript);
+        }
+
+        return manuscript.get();
```
Comment (stefan-nistor) [inline Backend/src/main/java/ro/uaic/info/romandec/services/ManuscriptService.java:87]:
```suggestion
        return manuscriptRepository.getManuscriptByIdAndNameAndUserId(
                request.getManuscriptId(),
                request.getName(),
                userId).orElseThrow(() -> new NoAvailableDataForGivenInput("No manuscript found for id:" + request.getManuscriptId()));
```

File: Backend/src/main/java/ro/uaic/info/romandec/controllers/ProfileController.java
Comment (stefan-nistor) [inline Backend/src/main/java/ro/uaic/info/romandec/controllers/ProfileController.java:1]: You should remove try-catch blocks from Controller layer. The exception handling should take place in the Service layer. The thrown exceptions are handled in ExceptionHandlerAdvice

File: Backend/src/main/java/ro/uaic/info/romandec/services/ManuscriptService.java [hunk 5]
```diff
@@ -2,44 +2,168 @@
...
+        for (Manuscript m : usersManuscripts) {
+            response.add(ManuscriptPreviewResponse
+                    .builder()
+                    .manuscriptId(m.getId())
+                    .name(m.getName())
+                    .build());
+        }
+
+        return response;
+    }
```
Comment (stefan-nistor) [inline Backend/src/main/java/ro/uaic/info/romandec/services/ManuscriptService.java:141]:
```suggestion
    public List<ManuscriptPreviewResponse> getAllUsersManuscripts(UUID userId)
        throws NoAvailableDataForGivenInput {
    List<Manuscript> usersManuscripts = manuscriptRepository.getAllByUserId(userId);

    if (usersManuscripts == null) {
        throw new NoAvailableDataForGivenInput("No available manuscripts for this user.");
    }

    return usersManuscripts.stream()
            .map(m -> ManuscriptPreviewResponse.builder()
                    .manuscriptId(m.getId())
                    .name(m.getName())
                    .build())
            .collect(Collectors.toList());
}
```

File: Backend/src/main/java/ro/uaic/info/romandec/services/ManuscriptService.java [hunk 6]
```diff
@@ -2,44 +2,168 @@
...
+            manuscriptRepository.save(manuscript);
+
+            Files.writeString(filePath, decipheredText,
+                    StandardOpenOption.CREATE, StandardOpenOption.APPEND);
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeException("Error at creating the file with the deciphered text");
+        }
+    }
```
Comment (stefan-nistor) [inline Backend/src/main/java/ro/uaic/info/romandec/services/ManuscriptService.java:118]: You can extract some methods to improve code readability and reduce cognitive complexity

```suggestion
    public void saveAnnotatorDecipheredManuscript(String originalImageFilename, String decipheredText)
        throws InvalidDataException, IOException {
    String filenameWithoutExtension = extractFilenameWithoutExtension(originalImageFilename);
    Path imageDirectory = Paths.get(databaseDirectory, filenameWithoutExtension);

    verifyImageDirectory(imageDirectory);

    Path filePath = resolveFilePath(imageDirectory, originalImageFilename);
    updateManuscript(imageDirectory, originalImageFilename, filePath);
    Files.writeString(filePath, decipheredText, StandardOpenOption.CREATE, StandardOpenOption.APPEND);
}

private void verifyImageDirectory(Path imageDirectory) throws InvalidDataException {
    if (!Files.exists(imageDirectory)) {
        throw new InvalidDataException("This image does not a directory associated");
    }
}

private Path resolveFilePath(Path imageDirectory, String originalImageFilename) {
    return imageDirectory.resolve(Objects.requireNonNull(getDecipheredTranscriptFilename(originalImageFilename)));
}

private void updateManuscript(Path imageDirectory, String originalImageFilename, Path filePath) {
    Path imageCompletePath = imageDirectory.resolve(originalImageFilename);
    Manuscript manuscript = manuscriptRepository.getManuscriptByPathToImage(imageCompletePath.toAbsolutePath().toString());
    manuscript.setPathToDecipheredText(filePath.toAbsolutePath().toString());
    manuscriptRepository.save(manuscript);
}
```

**Review by stefan-nistor — CHANGES_REQUESTED** [review]

File: Backend/src/main/java/ro/uaic/info/romandec/repository/ManuscriptRepository.java [hunk 7]
```diff
@@ -4,12 +4,20 @@
 import org.springframework.data.jpa.repository.Query;
 import ro.uaic.info.romandec.models.Manuscript;
 
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
 public interface ManuscriptRepository extends JpaRepository<Manuscript,Integer> {
```
Comment (stefan-nistor) [inline Backend/src/main/java/ro/uaic/info/romandec/repository/ManuscriptRepository.java:11]: Annotate with `@Repository`
```suggestion
@Repository
public interface ManuscriptRepository extends JpaRepository<Manuscript,Integer> {
```

**Review by stefan-nistor — CHANGES_REQUESTED** [review]

File: Backend/src/main/java/ro/uaic/info/romandec/repository/ManuscriptRepository.java (hunk 7 above)
Comment (stefan-nistor) [inline Backend/src/main/java/ro/uaic/info/romandec/repository/ManuscriptRepository.java:11]: Please annotate with `@Repository`
```suggestion
@Repository
public interface ManuscriptRepository extends JpaRepository<Manuscript,Integer> {
```

**Review by stefan-nistor — CHANGES_REQUESTED** [review]
Some changes are needed regarding ExceptionHandling. We should not extend directly `Exception` class. Made a new branch for diff checking - a5406c4

**Review by stefan-nistor — APPROVED** [review]

**Review by StefanRoman1 — APPROVED** [review]

**Pull request merged**